shell.go

```diff
@@ -371,6 +374,52 @@ func (s *Shell) FindPeer(peer string) (*PeerInfo, error) {
 	return &str.Responses[0], nil
 }
 
+func (s *Shell) FindProvs(ctx context.Context, cid string) (<-chan pstore.PeerInfo, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	resp, err := s.newRequest("dht/findprovs", cid).Send(s.httpcli)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Error != nil {
+		return nil, resp.Error
+	}
+
+	outchan := make(chan pstore.PeerInfo, 4)
+
+	go func() {
+		defer close(outchan)
+		defer cancel()
+
+		var n notif.QueryEvent
+		decoder := json.NewDecoder(resp.Output)
+		for {
+			err := decoder.Decode(&n)
+			if err != nil {
+				return
+			}
+
+			if n.Type == notif.Provider {
+				for _, p := range n.Responses {
+					select {
+					case outchan <- *p:
+					case <-ctx.Done():
+						return
+					}
+				}
+			}
+		}
+	}()
+
+	go func() {
+		<-ctx.Done()
+		resp.Close()
```

Comment:
Wait, if `NewDecoder` hangs, then it's going to hang until someone cancels the context, because `resp.Close()` is not called until then, and the context won't be cancelled because `NewDecoder` is hanging ?